File: audience/src/main/java/br/com/concretesolutions/audience/core/actor/FragmentChoreography.java
package br.com.concretesolutions.audience.core.actor;


import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;

import br.com.concretesolutions.audience.core.Director;

public class FragmentChoreography extends FragmentManager.FragmentLifecycleCallbacks {

    @Override
    public void onFragmentCreated(FragmentManager fm, Fragment f, Bundle savedInstanceState) {
        super.onFragmentCreated(fm, f, savedInstanceState);

        if (!isActor(f))
            return;

        Director.actorRegistry().enroll((Actor) f, savedInstanceState);
    }

    @Override
    public void onFragmentSaveInstanceState(FragmentManager fm, Fragment f, Bundle outState) {
        super.onFragmentSaveInstanceState(fm, f, outState);

        if (!isActor(f))
            return;

        if (Director.isInConfigurationChange()) {
            Director.actorRegistry().take5((Actor) f, outState);
        }
    }

    @Override
    public void onFragmentDestroyed(FragmentManager fm, Fragment f) {
        super.onFragmentDestroyed(fm, f);

        if (!isActor(f))
            return;

        Director.play().stop(Director.actorRef((Actor) f));
    }

    private boolean isActor(Fragment f) {
        return f instanceof Actor;
    }
}

File: audience-core/src/main/kotlin/io/github/victorolinasc/audience/core/actor/FragmentChoreography.kt
package io.github.victorolinasc.audience.core.actor


import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager

import io.github.victorolinasc.audience.core.Director

class FragmentChoreography : FragmentManager.FragmentLifecycleCallbacks() {

    override fun onFragmentCreated(fm: FragmentManager?, f: Fragment?, savedInstanceState: Bundle?) {
        super.onFragmentCreated(fm, f, savedInstanceState)

        if (!isActor(f))
            return

        Director.system.actorRegistry.enroll((f as Actor?)!!, savedInstanceState)
    }

    override fun onFragmentSaveInstanceState(fm: FragmentManager?, f: Fragment?, outState: Bundle?) {
        super.onFragmentSaveInstanceState(fm, f, outState)

        if (!isActor(f))
            return

        if (Director.system.isInConfigurationChange) {
            Director.system.actorRegistry.take5(f as Actor, outState!!)
        }
    }

    override fun onFragmentDestroyed(fm: FragmentManager?, f: Fragment?) {
        super.onFragmentDestroyed(fm, f)

        if (!isActor(f))
            return

        Director.system.stop(Director.actorRef((f as Actor)))
    }

    private fun isActor(f: Fragment?) = f is Actor
}
